hibernate-core/src/main/java/org/hibernate/boot/model/internal/EmbeddableBinder.java

```diff
@@ -852,11 +853,38 @@ private static void processIdClassElements(
 					//the annotation overriding will be dealt with by a mechanism similar to @MapsId
 					continue;
 				}
+				if ( !hasCompatibleType( idClassPropertyData.getTypeName(), entityPropertyData.getTypeName() ) ) {
+					throw new AnnotationException(
+							"Property '" + propertyName + "' in @IdClass '" + idClassPropertyData.getDeclaringClass().getName()
+									+ "' doesn't match type in entity class '" + baseInferredData.getPropertyType().getName()
+									+ "' (expected '" + entityPropertyData.getTypeName() + "' but was '" + idClassPropertyData.getTypeName() + "')"
+					);
+				}
 			}
 			classElements.set( i, entityPropertyData );  //this works since they are in the same order
 		}
 	}
 
+	private static boolean hasCompatibleType(String typeNameInIdClass, String typeNameInEntityClass) {
+		return typeNameInIdClass.equals( typeNameInEntityClass )
+				|| canonicalize( typeNameInIdClass ).equals( typeNameInEntityClass )
+				|| typeNameInIdClass.equals( canonicalize( typeNameInEntityClass ) );
```

Comment:
I still think this code is equivalent to `canonicalize( typeNameInIdClass ).equals( canonicalize( typeNameInEntityClass ) )`. Is that not the case?

Reply:
Yes, but that is less efficient, `typeNameInIdClass.equals( typeNameInEntityClass )` will be hit in most cases.

Reply:
> Yes, but that is less efficient, `typeNameInIdClass.equals( typeNameInEntityClass )` will be hit in most cases.

@gavinking WDYT?